File: src/main/resources/static/python/descending/QuickSortDescending.py
import random
import time
import csv
import os

def quick_sort(arr):
    """Sorts an array in-place using the Quick Sort algorithm."""
    def partition(low, high):
        pivot = arr[high]
        i = low - 1
        for j in range(low, high):
            if arr[j] > pivot:
                i += 1
                arr[i], arr[j] = arr[j], arr[i]
        arr[i + 1], arr[high] = arr[high], arr[i + 1]
        return i + 1

    def quick_sort_recursive(low, high):
        if low < high:
            pi = partition(low, high)
            quick_sort_recursive(low, pi - 1)
            quick_sort_recursive(pi + 1, high)

    quick_sort_recursive(0, len(arr) - 1)

def write_to_file(filename, arr):
    with open(filename, 'w', newline='') as file:
        writer = csv.writer(file)
        writer.writerow(arr)

def read_from_file(filename):
    if not os.path.exists(filename):
        print(f'File not found: {filename}')
        return []
    with open(filename, 'r') as file:
        reader = csv.reader(file)
        return list(map(int, next(reader)))

def main():
    # Get the array size dynamically from user input
    try:
        n = int(input("Enter the size of the array: "))  # Get array size from user input
        if n <= 0:
            raise ValueError
    except ValueError:
        print("Array size must be a positive integer.")
        os.sys.exit(1)

    # Navigate two levels up to reach the static directory
    base_path = os.path.abspath(os.path.join(os.path.dirname(__file__), '../../'))
    original_file = os.path.join(base_path, 'original_array.csv')
    sorted_file = os.path.join(base_path, 'sorted_result.csv')

    if os.path.exists(original_file):
        # print('Original array file found. Reading from file...')
        arr = read_from_file(original_file)
    else:
        # print('Original array file not found. Creating new array...')
        arr = [random.randint(1, 100) for _ in range(n)]
        write_to_file(original_file, arr)

    start_time = time.time()

    quick_sort(arr)

    write_to_file(sorted_file, arr)

    end_time = time.time()
    total_time = end_time - start_time
    
    # # Output results with HTML-style links
    # print(f'Original array file: <a href="/static/original_array.csv">original_array.csv</a>')
    # print(f'Sorted array file: <a href="/static/sorted_result.csv">sorted_result.csv</a>')
    print(f'Time taken: {total_time:.3f} seconds.')

if __name__ == "__main__":
    main()

File: src/main/resources/static/javascript/descending/QuickSortDescending.js
const fs = require('fs');
const path = require('path');

// Quick Sort Algorithm for Descending Order
function quickSort(arr) {
    if (arr.length <= 1) return arr;
    const pivot = arr[Math.floor(arr.length / 2)];
    const left = arr.filter(x => x > pivot); // Change to > for descending order
    const middle = arr.filter(x => x === pivot);
    const right = arr.filter(x => x < pivot); // Change to < for descending order
    return [...quickSort(left), ...middle, ...quickSort(right)];
}

// Write array to a file
function writeToFile(filename, arr) {
    const data = arr.join(',') + '\n';
    fs.writeFileSync(filename, data);
}

// Read array from a file
function readFromFile(filename) {
    if (!fs.existsSync(filename)) {
        console.error('File not found:', filename);
        return [];
    }
    const content = fs.readFileSync(filename, 'utf-8');
    return content.split(',').map(num => parseInt(num.trim(), 10));
}

// Main function
function main() {
    if (process.argv.length < 3) {
        console.error('Usage: node script.js <array-size>');
        process.exit(1);
    }

    const n = parseInt(process.argv[2], 10); // Get array size from command-line argument
    if (isNaN(n) || n <= 0) {
        console.error('Array size must be a positive integer.');
        process.exit(1);
    }

    const basePath = path.join(__dirname, '../../'); // Go up two levels to reach the static directory
    const originalFile = path.join(basePath, 'original_array.csv');
    const sortedFile = path.join(basePath, 'sorted_result.csv');

    // Always create a new array based on user input
    let arr = Array.from({ length: n }, () => Math.floor(Math.random() * 100) + 1);
    writeToFile(originalFile, arr);

    const startTime = Date.now();

    arr = quickSort(arr);

    writeToFile(sortedFile, arr);

    const endTime = Date.now();
    const totalTime = (endTime - startTime) / 1000;

    // // Output results with HTML-style links
    // console.log(`Original array file: <a href="/static/original_array.csv">original_array.csv</a>`);
    // console.log(`Sorted array file: <a href="/static/sorted_result.csv">sorted_result.csv</a>`);
    console.log(`Time taken: ${totalTime} seconds.`);
}

main();
